src/boot/efi/chid.c

```diff
@@ -0,0 +1,385 @@
+                                L"&",
+                                info->family,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 11:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->family);
+                break;
+        case 13:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                5,
+                                info->manufacturer,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 14:
+                status = hash_strings_sha1(
+                                &hash, (uint8_t *) &namespace, sizeof(namespace), 1, info->manufacturer);
+                break;
+        default:
+                return EFI_SUCCESS; /* Just keep empty to prevent match. */
+        }
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        memcpy(chid, hash, sizeof(*chid));
+
+        /* Convert the resulting CHID back to little-endian: */
+        chid->Data1 = __builtin_bswap32(chid->Data1);
+        chid->Data2 = __builtin_bswap16(chid->Data2);
+        chid->Data3 = __builtin_bswap16(chid->Data3);
+
+        /* set specific bits according to RFC4122 Section 4.1.3 */
+        chid->Data3 = (chid->Data3 & 0x0fff) | (5 << 12);
+        chid->Data4[0] = (chid->Data4[0] & 0x3f) | 0x80;
+
+        return EFI_SUCCESS;
+}
+
+/**
+ * populate_board_hwids() - Read board SMBIOS and produce an array of CHID values.
+ * @hwids:  Pointer to an array of 12 chids to be filled.
+ */
+static EFI_STATUS populate_board_hwids(EFI_GUID *hwids) {
+        EFI_STATUS status;
+        _cleanup_(smbios_info_clear) SmbiosInfo info = {};
+        int i;
+
+        if (!hwids)
+                return EFI_INVALID_PARAMETER;
+
+        status = populate_smbios_info(&info);
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        for (i = 0; i < 15; ++i) {
+                status = get_chid(&info, i, &hwids[i]);
+                if (EFI_STATUS_IS_ERROR(status))
+                        goto exit;
+        }
+
+exit:
+        return status;
+}
+
+typedef struct Device {
+        const char16_t name[128];
+        const char compatible[128];
+        EFI_GUID hwids[32];
```

Comment:
where do these sizes come from? they really should have some explanation

Reply:
`compatible` size is from [Linux](https://github.com/torvalds/linux/blob/master/include/linux/mod_devicetable.h#L285), others are more or less arbitrary. I could probably drop the `name` altogether as it's not used yet or make it `char` and log it to the user (it is useful information as a basic sanity-check, though I'm not sure how it should be done). `hwids` should probably have the length of 15 or 16 since there usually 15 values in `fwupdtool hwids` output. I initially thought that it might be worth it to have more i.e. if there is a different SKU with basically the same hardware, but it probably should just be in a separate entry.

Reply:
all i care about is that this code here comes with comments where this comes from.